domains.c

```diff
@@ -787,29 +800,38 @@ static void domain_calculate_abs_scaling(const char *path, void *opaque)
  *  Returns a non-zero value if this slot is occuped by a left-over domain.
  *
  */
-static int slot_occupied_by_dead_domain(int slot)
+static int slot_occupied_by_dead_domain(int slot, int domid)
 {
     int dying, ret;
     xc_dominfo_t info;
     char * reported_domid;
 
     struct domain * d;
-
     //If we've been passed the special slot -1, vacuously return false;
     //this is an indication that a domain has no real slot (and thus can't
     //occupy a slot).
     if(slot == -1) {
         return 0;
     }
 
-    d = domain_with_slot(slot);
+    d = domain_with_slot_and_not_domid(slot, domid);
+
+    //Some other domain, likely dead, occupies this slot, deal with it
+    if (d) {
+        ret = xc_domain_getinfo(xc_handle, d->domid, 1, &info);
+        //Xen couldn't find the domain or the domain info it returned is not
+        //the domain we asked for, so it has NO idea about it.
+        if(ret != 1 || (int)info.domid != d->domid) {
```

Comment:
In case I am not the only one to squint at the above:
`xc_domain_getinfo` second parameter is `first_domid`, used as a start to iterate on a number (3rd arg) of contiguous domids. Each domid is passed to `domctl` `getdomaininfo` and breaks the loop if the `domctl` returns an error, **but** `domctl` `getdomaininfo` will return the info of the **first existing domid higher than** the one passed as argument[[1](https://xenbits.xen.org/gitweb/?p=xen.git;a=blob;f=xen/common/domctl.c;h=bade9a63b18d33a8306da297720b1428d6e47ae6;hb=HEAD#l746)].

Grepping around there is another occurrence of this in `Surfman`: https://github.com/OpenXT/surfman/blob/master/surfman/src/domain.c#L38

Nice find!